### PR TITLE
Check for Kool updates when starting containers.

### PR DESCRIPTION
**Pull request by nkabz**
Added logic to check if theres a new version for Kool when starting containers. This is the first iteration, there is still work to do (mainly tests and refactors as seem fit). As @fabriciojs suggested, using a decorator to wrap the start function and execute the check.

Question: should this check be limited for a fixed number of times? Say, every 5 times a user uses `kool start`?

Would be nice to get advise from @danielsuguimoto on how to procede with tests in this particupar case :smile:

This PR closes ticket #132

### CODE REVIEW

File: cmd/kool_upgrade_available_checker.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+	KoolService
+
+	updater updater.Updater
+}
+
+// UpdateWrapper wraps the service with checker logic
+func UpdateWrapper(service KoolService) *UpdateAwareService {
+	return &UpdateAwareService{
+		service,
+		&updater.DefaultUpdater{RootCommand: rootCmd},
```
Comment (danielsuguimoto) [inline cmd/kool_upgrade_available_checker.go:20]: We need to refactor it, so here we can receive the `DefaultUpdater` or the `FakeUpdater`

**Comment by nkabz** [comment]
@danielsuguimoto Tests added and refactor made, waiting on review for some more CR changes!

**Review by danielsuguimoto — CHANGES_REQUESTED** [review]
LGTM! I would add just one more test case, checking that we are not checking for new updates on non-tty environments

image

File: cmd/kool_upgrade_available_checker.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	}
+}
+
+// Execute runs the check logic and proxies to original service
+func (u *UpdateAwareService) Execute(args []string) (err error) {
+	if !u.KoolService.IsTerminal() {
+		if err = u.KoolService.Execute(args); err != nil {
+			return err
+		}
+		return
```
Comment (fabriciojs) [inline cmd/kool_upgrade_available_checker.go:30]:
```suggestion
		err = u.KoolService.Execute(args)
		return
```

Does the same.

File: cmd/kool_upgrade_available_checker.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	go u.updater.CheckForUpdates(u.updater.GetCurrentVersion(), ch)
+
+	if err = u.KoolService.Execute(args); err != nil {
+		return err
+	}
+
+	select {
+	case update := <-ch:
+		if update {
+			defer u.KoolService.Warning("Theres a new Kool Version available! Run kool self-update to update!")
```
Comment (fabriciojs) [inline cmd/kool_upgrade_available_checker.go:44]:
```suggestion
			defer u.KoolService.Warning("There's a new version available! Run kool self-update to upgrade!")
```

File: cmd/kool_upgrade_available_checker.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+		}
+	case <-time.After(300 * time.Millisecond):
```
Comment (fabriciojs) [inline cmd/kool_upgrade_available_checker.go:46]: Maybe we can wait up to 1 second?

File: cmd/updater/updater.go [hunk 5]
```diff
@@ -42,3 +43,25 @@ func (u *DefaultUpdater) Update(currentVersion semver.Version) (updatedVersion s
...
+// CheckForUpdates checks if there is a new version
+func (u *DefaultUpdater) CheckForUpdates(current semver.Version, ch chan bool) {
+	var (
+		latest       *selfupdate.Release
+		isNewVersion = false
+		err          error
+	)
+
+	if latest, _, err = selfupdate.DetectLatest("kool-dev/kool"); err != nil {
+		ch <- isNewVersion
```
Comment (fabriciojs) [inline cmd/updater/updater.go:56]:
```suggestion
		chHasNewVersion <- false
```

File: cmd/updater/updater.go [hunk 6]
```diff
@@ -42,3 +43,25 @@ func (u *DefaultUpdater) Update(currentVersion semver.Version) (updatedVersion s
...
+		return
+	}
+
+	if !latest.Version.Equals(current) {
+		isNewVersion = true
```
Comment (fabriciojs) [inline cmd/updater/updater.go:61]: chHasNewVersion <- true

File: cmd/updater/updater.go [hunk 7]
```diff
@@ -42,3 +43,25 @@ func (u *DefaultUpdater) Update(currentVersion semver.Version) (updatedVersion s
 	updatedVersion = latest.Version
 	return
 }
+
+// CheckForUpdates checks if there is a new version
+func (u *DefaultUpdater) CheckForUpdates(current semver.Version, ch chan bool) {
```
Comment (fabriciojs) [inline cmd/updater/updater.go:48]:
```suggestion
func (u *DefaultUpdater) CheckForUpdates(current semver.Version, chHasNewVersion chan bool) {
```

File: cmd/updater/updater.go [hunk 8]
```diff
@@ -42,3 +43,25 @@ func (u *DefaultUpdater) Update(currentVersion semver.Version) (updatedVersion s
...
+	var (
+		latest       *selfupdate.Release
+		isNewVersion = false
```
Comment (fabriciojs) [inline cmd/updater/updater.go:51]:
```suggestion
```

File: cmd/updater/updater.go [hunk 9]
```diff
@@ -42,3 +43,25 @@ func (u *DefaultUpdater) Update(currentVersion semver.Version) (updatedVersion s
...
+	if latest, _, err = selfupdate.DetectLatest("kool-dev/kool"); err != nil {
+		ch <- isNewVersion
+		return
+	}
+
+	if !latest.Version.Equals(current) {
+		isNewVersion = true
+	}
+
+	ch <- isNewVersion
```
Comment (fabriciojs) [inline cmd/updater/updater.go:64]:
```suggestion
```

**Review by fabriciojs — CHANGES_REQUESTED** [review]
@nkabz great!

Added a couple minor suggestions. Also +1 for Jun's remarks on testing to asset we do not update when not under TTY environment.

**Review by fabriciojs — APPROVED** [review]

**Review by danielsuguimoto — APPROVED** [review]
@nkabz Nice! Thx!

**Pull request merged**